Refactor net-vm

mbssrc opened the pull request:
<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of changes

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

Refactoring of netvmExtraModules:
  - remove netvm extra modules file and use pci passthrough module
  - create wifi config service
  - refactor host names to module to allow query ips
  - create reference service module + passthrough to VMs
  - moved dendrite as first reference service

## Checklist for things done

<!-- Please check, [X], to all that applies. Leave [ ] if an item does not apply but you have considered the check list item. Note that all of these are not hard requirements. They serve information to reviewers. When you fill the checklist, you indicate to reviewers you appreciate their work. -->

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [ ] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing

<!--
How this was tested by the author? How is this supposed to be tested
by people doing system testing?
-->

### Testing done 
- Build and run on x1
- checked `smcroute` service is running in net-vm 
- build with reference service disabled


### Testing required
- verify build on all platforms
- build and test on x1
- verify dendrite-pinecone + element works
- verify build works without dendrite-pinecone
- verify build works without reference services

brianmcgillion commented on targets/lenovo-x1/everything.nix:
```diff
@@ -79,9 +76,8 @@
-                    extraModules = [
-                      config.ghaf.hardware.passthrough.netvmPCIPassthroughModule
-                    ];
+                    dendrite = true;
```
netvm.extraModules = [somewhere.dendrite.module];

mbssrc replied:
done

brianmcgillion commented on modules/microvm/virtualization/microvm/modules.nix:
```diff
@@ -0,0 +1,122 @@
+  cfgPassthru = config.ghaf.hardware.passthrough;
+
+  # Dendrite netvm configuration
+  dendriteModule = [
+    {
+      config.ghaf.services.dendrite-pinecone = let
+        externalNic = let
+          firstPciWifiDevice = head config.ghaf.hardware.definition.network.pciDevices;
+        in "${firstPciWifiDevice.name}";
+
+        internalNic = let
+          vmNetworking = import ./common/vm-networking.nix {
+            inherit config;
+            inherit lib;
+            vmName = "net-vm";
+            inherit (config.microvm.net-vm) macAddress;
+            internalIP = 1;
+          };
+        in "${head vmNetworking.networking.nat.internalInterfaces}";
+
+        getElementVmEntry = filter (x: x.name == "element-vm") config.ghaf.networking.hosts.entries;
+        serverIpAddr = head (map (x: x.ip) getElementVmEntry);
+      in {
+        enable = true;
+        firewallConfig = true;
+        inherit externalNic;
+        inherit internalNic;
+        inherit serverIpAddr;
+      };
+    }
+  ];
```
as a module that is not here, but under reference-apps/reference-services/ anywhere-else. It is not part of the ghaf framework. It is needed by one application in one specific configuration

mbssrc replied:
done

mbssrc commented on modules/common/networking/hosts.nix:
```diff
@@ -58,17 +78,55 @@
+    };
+
+    # Generate hosts file
+    environment.etc.hosts = lib.mkForce {
```
Yes - the networking definition is being re-worked as we speak and this should be addressed in the next iteration.

brianmcgillion replied:
Being worked on with Tervis

milva-unikie commented:
Hi! Can you please provide more detailed instructions on how these should be tested?

> - verify dendrite-pinecone + element works
> - verify build works without dendrite-pinecone
> - verify build works without reference services

- How should it be confirmed that dendrite-pinecone + element works?
- How should dendrite-pinecone and reference services be disabled?
- What are the expected changes from disabling them?

mbssrc committed:
Netvm refactoring

- remove netvm extra modules file
- centralize targets module processing
- create wifi config service
- create dendrite profile and automatic inclusion
- refactor host names to module to allow query ips

Signed-off-by: Manuel Bluhm <manuel@ssrc.tii.ae>

 WIP:

- guarding module generator for x86
- removing firmwares from netvm

Signed-off-by: Manuel Bluhm <manuel@ssrc.tii.ae>

 Introduce reference services

- create reference services that are not part of Ghaf
- move dendrite
- update modules.nix as interface to different vms

Signed-off-by: Manuel Bluhm <manuel@ssrc.tii.ae>

Adding HW namespace for VMs and changing reference services
- simplified reference service and passthrough
- added hardware.definition namespace to common, and added population of values in accordance with passthrough modules.
- removed graphics.hardware - no longer needed

Signed-off-by: Manuel Bluhm <manuel@ssrc.tii.ae>

mbssrc commented:
Hey, absolutely, more details below.
Note that the last version is https://github.com/tiiuae/ghaf/commit/ca8ceeb28f404b69e70b3e414a5ee60f054cac14 .

**- How should it be confirmed that dendrite-pinecone + element works?**
Can you test it similar to the original PR (https://github.com/tiiuae/ghaf/pull/533)?
If you cannot test this end-to-end, at least the following should be verified:

- `smcroute.service` is running in the netvm
- `dendrite-pinecone.service` is running in the element-vm

**- How should dendrite-pinecone and reference services be disabled?**
In everything.nix, there is the configuration:
```
reference.services = {
    enable = true;
    dendrite = true;
  };
```
Setting these options to `false` respectively should do the trick. 

**- What are the expected changes from disabling them?**
1. Build on x1 works fine
2. Aforementioned services are not present/running if either of them are turned off

johannarautanen commented:
Check with commit: ca8ceeb28f404b69e70b3e414a5ee60f054cac14

**NVIDIA Jetson AGX Orin native, booted Orin AGX from eMMC**

- ci-test-automation tests were ok
- all apps launched manually ok
- All manual regression test ok


**NVIDIA Jetson NX Orin native, booted Orin NX from USB flash drive**

- ci-test-automation tests were ok
- all apps launched manually ok
- All manual regression test ok

barnabakos commented:
with commits up to: ca8ceeb28f404b69e70b3e414a5ee60f054cac14

platform: **Lenovo X1 Carbon**
- **all build variations:**
  - ci-test-automation suite: ok
  - all apps start
  - ad-hoc tesing: no new findings
- **default commit** (both dendrite-pinecone & reference services enabled):
  - smcroute & dendrite services are running; logs are shown
- only **dendrite-pinecone** disabled:
  - smcroute & dendrite services are not running & not available
- only **reference services** disabled:
  - smcroute is not running & not available
  - **dendrite is running; logs are shown**
- both **dendrite-pinecone & reference services** disabled:
  - smcroute & dendrite services are not running & not available

platform: **NVIDIA Jetson AGX Orin with demo apps; cross-compile**
- ci-test-automation suite: ok (SP-4729 - firefox doesn't start -> known issue)
- all apps start
- ad-hoc tesing: no new findings
